### PR TITLE
Creates an audit log across services

### PR DESCRIPTION
**Pull request by fhats**
This commit creates an "audit log" of administrative actions taken on services
and clusters. The goal of this log is to provide a central view of actions
being taken to services across the platform. Currently this information is both
incompletely and inconsistently logged, and the information that is logged is
distributed across every service's individual logs, which is challenging and
slow to collate.

This change also renames the setting for the 'file' log writer that was
previously named 'delimeter' to 'delimiter', because spelling.

### Test Plan

I've tested this change by:

* running the unit tests (spoiler: they passed)
* running the itests (spoiler: they passed)
* poking around with the [example cluster](https://paasta.readthedocs.io/en/latest/installation/example_cluster.html)

The example cluster produced positive results for the `cook-image`, `emergency-stop/start/restart`, `mark-for-deployment`, `pause-service-autoscaler`, `push-to-registry`, `rerun`, `start`, `stop`, `restart`, `rollback`, and `boost` commands (even though the cluster boost command is disabled in the example cluster). This change contains support for audit logging in the `autoscale`, `generate-pipeline`, and `secret add/update` commands, but I was unable to validate those commands on the example cluster. I've made a note to test those specifically on mesosstage. I've also not yet run this code configured to log to scribe (any hints on how to do that prior to releasing to mesosstage?).

Here's what the audit log looks like for the actions I was able to validate:

```
{"action": "push-to-registry", "action_details": {"commit": "9cdf957d3fbbb61fa17388adb2d8e282446271fc"}, "cluster": "N/A", "host": "e93514543345", "instance": "N/A", "service": "hello-world", "timestamp": "2019-01-18T19:22:57.826991", "user": "root"}
```

```
{"action": "mark-for-deployment", "action_details": {"commit": "9cdf957d3fbbb61fa17388adb2d8e282446271fc", "deploy_group": "testcluster.everything"}, "cluster": "N/A", "host": "e93514543345", "instance": "N/A", "service": "hello-world", "timestamp": "2019-01-18T19:23:06.415982", "user": "root”}
```

```
{"action": "cook-image", "action_details": {"tag": "paasta-cook-image-hello-world-root"}, "cluster": "N/A", "host": "e93514543345", "instance": "N/A", "service": "hello-world", "timestamp": "2019-01-18T19:27:55.220279", "user": "root"}
```

```
{"action": "emergency-stop", "action_details": {}, "cluster": "testcluster", "host": "c906b1737af0", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:58:16.022126", "user": "root"}
{"action": "emergency-start", "action_details": {}, "cluster": "testcluster", "host": "c906b1737af0", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:58:44.572353", "user": "root"}
{"action": "emergency-restart", "action_details": {}, "cluster": "testcluster", "host": "c906b1737af0", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:59:14.576234", "user": "root"}
```

```
{"action": "pause-service-autoscaler", "action_details": {"duration": 30}, "cluster": "testcluster", "host": "e93514543345", "instance": "N/A", "service": null, "timestamp": "2019-01-18T19:33:05.925792", "user": "root"}
{"action": "resume-service-autoscaler", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "N/A", "service": null, "timestamp": "2019-01-18T19:33:14.844322", "user": "root”}
```

```
{"action": "chronos-rerun", "action_details": {"execution_date": "2019-01-18T19:35:06", "rerun_type": null}, "cluster": "testcluster", "host": "e93514543345", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:35:07.311133", "user": "root”}
```

```
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "main", "service": "hello-world", "timestamp": "2019-01-18T19:37:52.351385", "user": "root"}
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:37:52.468566", "user": "root"}
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "remote", "service": "hello-world", "timestamp": "2019-01-18T19:37:52.592409", "user": "root”}    
```

```
{"action": "stop", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "main", "service": "hello-world", "timestamp": "2019-01-18T19:38:31.321584", "user": "root"}
{"action": "stop", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:38:31.440025", "user": "root"}
{"action": "stop", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "remote", "service": "hello-world", "timestamp": "2019-01-18T19:38:31.556309", "user": "root”}
```

```
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "main", "service": "hello-world", "timestamp": "2019-01-18T19:39:00.673805", "user": "root"}
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "test_chronos", "service": "hello-world", "timestamp": "2019-01-18T19:39:00.788387", "user": "root"}
{"action": "start", "action_details": {}, "cluster": "testcluster", "host": "e93514543345", "instance": "remote", "service": "hello-world", "timestamp": "2019-01-18T19:39:00.904008", "user": "root”}
```

```
{"action": "mark-for-deployment", "action_details": {"commit": "9cdf957d3fbbb61fa17388adb2d8e282446271fc", "deploy_group": "testcluster.everything"}, "cluster": "N/A", "host": "e93514543345", "instance": "N/A", "service": "hello-world", "timestamp": "2019-01-18T19:39:58.197461", "user": "root”}
```

```
{"action": "cluster-boost", "action_details": {"boost": 1.5, "boost_action": "set", "duration": 10, "override": false, "pool": "default"}, "cluster": "testcluster", "host": "e93514543345", "instance": "N/A", "service": null, "timestamp": "2019-01-18T19:41:02.163337", "user": "root"}
```

### CODE REVIEW

File: paasta_tools/utils.py [hunk 1]
```diff
@@ -1214,6 +1297,28 @@ def log(
...
+        formatted_line = format_audit_log_line(
+            user=user,
+            host=host,
+            action=action,
+            action_details=action_details,
+            service=service,
+            cluster=cluster,
+            instance=instance,
+        )
+        self.clog.log_line(log_name, formatted_line)
```
Comment (EvanKrall) [inline paasta_tools/utils.py]: How do you feel about _also_ logging to the service's log? I think a common use case will be to look at these audit logs for a particular service, e.g. with `paasta logs`.

File: paasta_tools/utils.py [hunk 2]
```diff
@@ -1273,39 +1390,66 @@ def format_path(self, service: str, component: str, level: str, cluster: str, in
...
 
-        to_write = "{}{}".format(
-            format_log_line(level, cluster, service, instance, component, line),
-            self.line_delimeter,
-        )
-
         try:
             with io.FileIO(path, mode=self.mode, closefd=True) as f:
                 with self.maybe_flock(f):
                     # remove type ignore comment below once https://github.com/python/typeshed/pull/1541 is merged.
```
Comment (EvanKrall) [inline paasta_tools/utils.py]: While you're here, do you mind removing this line and the `# type: ignore` comment below?

File: tests/cli/test_cmds_secret.py [hunk 3]
```diff
@@ -135,6 +137,14 @@ def test_paasta_secret():
             secret_name='theonering',
             plaintext=mock_get_plaintext_input.return_value,
         )
+        mock_log_audit.assert_called_with(
+            action='add-secret',
+            action_details={
+                'secret_name': 'theonering',
```
Comment (EvanKrall) [inline tests/cli/test_cmds_secret.py]: what does it mean to theoner

File: paasta_tools/utils.py (hunk 1 above)
Comment (fhats) [reply paasta_tools/utils.py]: decided against this for now, requires some editorial decisions to the other log functionality that i'd just as soon skip for the moment

**Pull request merged**